docs: Add channel configuration file design proposal

pavelzw opened the pull request:
I turned my discussions with @wolfv into a design proposal. Please take a look and check if all relevant use cases are covered.

pavelzw commented on docs/design_proposals/channel_config.md:
```diff
@@ -0,0 +1,176 @@
+5. `$HOME/rattler.toml`
+6. `$HOME/.rattler.toml`
+
+If you want to use a different configuration location, you can override it with the `RATTLER_CHANNEL_CONFIG` environment variable.
```
Maybe `RATTLER_CONFIG` instead of `RATTLER_CHANNEL_CONFIG` if we want to specify different behavior as well in this file at some point.

0xbe7a commented:
I think the current proposal for configuring channel sources using a combination of names, URLs, and mirrors introduces unnecessary complexity and could potentially confuse users. This approach complicates the understanding and management of package sources by requiring users to juggle between multiple layers of configuration: the channel name, the URL in lockfiles, and the actual URL used for fetching packages.

I think having just a direct name-to-URL mapping in the rattler.toml configuration file and including only the channel names in the lockfiles would be a lot simpler. This approach would make the lockfiles portable across different environments without tying them to specific source URLs. It removes the need for probing which mirror is available and avoids the redundancy and confusion of managing multiple sources for a single channel.

baszalmstra commented:
@pavelzw Thanks for starting this document! 👍 

I agree with @0xbe7a that the use of names, urls, and mirrors is quite confusing. But having a direct mapping would probably also mean you would always need a configuration file with some defaults. I thinking having a default way to resolve a name to a URL is also useful.

pavelzw commented:
> having a default way to resolve a name to a URL is also useful

Prepending `https://conda.anaconda.org/` if nothing else is specified would solve this issue.
But then, the URL shouldn't be in the lockfile imo, maybe something like https://github.com/prefix-dev/pixi/pull/931#issuecomment-1987152752 could be a solution.

baszalmstra commented:
> > having a default way to resolve a name to a URL is also useful
> 
> But then, the URL shouldn't be in the lockfile imo, maybe something like [#931 (comment)](https://github.com/prefix-dev/pixi/pull/931#issuecomment-1987152752) could be a solution.
Im really on the fence about that because you could easily get into a situation where if you dont share the same config as the author of the lock-file it wont install for you. Perhaps the mirror configuration should also be part of the lock-file??

pavelzw commented:
I get that, although then the lockfile won't be the same depending on what your `rattler.toml` is... Also, you probably don't want to leak corporate domain names to the outside world accidentally.

pavelzw commented:
How about this idea?

pixi.toml
```toml
channels = ["conda-forge", "my-channel"]

# this is the implicit case
[[channels]]
name = "conda-forge"
url = "https://conda.anaconda.org/conda-forge"

[[channels]]
name = "my-channel"
url = "https://repo.prefix.dev/my-channel"
```

pixi.lock
```yml
channels:
  - name: conda-forge
    url: https://conda.anaconda.org/conda-forge
  - name: my-channel
    url: https://repo.prefix.dev/my-channel
environments:
  default:
    channels:
    - conda-forge
    packages:
      linux-64:
      - conda: https://conda.anaconda.org/conda-forge/linux-64/_libgcc_mutex-0.1-conda_forge.tar.bz2
# ...
```

This makes the pixi.toml + pixi.lock self-contained

pavelzw commented:
I removed the `mirror` notion and replaced everything with `url` and added the option to specify the channel configuration in `pixi.toml`. This should hopefully address Bas's concerns as now, you can specify the `pixi.toml` to use private channels s.t. it still is sort-of self-contained and doesn't rely on `rattler.toml`.

@wolfv @baszalmstra @0xbe7a PTAL

pavelzw commented on docs/design_proposals/channel_config.md:
````diff
@@ -0,0 +1,216 @@
+# ...
+```
+
+### Multiple mirrors
````
I will probably not use this feature and I'm wondering how many people will... Is the performance difference that large?
In all corporate environments that I've worked in so far, you have a single mirror and in the outside world I probably just wouldn't bother setting up multiple mirrors and just use the default one.

If we don't support multiple mirrors for the same channel, it would make the config a bit easier to understand.

0xbe7a replied:
I also think multiple mirrors is a pretty rare use case and makes things a bit harder to understand, but I'm still on the fence about it. @wolfv do you have any specific use cases in mine where a list of mirrors would be required/beneficial?

wolfv replied:
We have a long standing goal of making conda-forge more independent from a single vendor. The goal is to achieve something akin to Linux distributions where you have multiple mirrors with telecom providers, universities, etc.

For now there are not many public mirrors (prefix.dev, ghcr.io and beta.mamba.pm, although that is not really up to date afaik). 

Ideally, with time, this would grow. It has the potential to also increase the reliability of the channels in general.

wolfv replied:
E.g. all the mirrors for Fedora: https://admin.fedoraproject.org/mirrormanager/mirrors/Fedora/39/x86_64

pavelzw commented on docs/design_proposals/channel_config.md:
````diff
@@ -0,0 +1,216 @@
+You can specify private channels in your `pixi.toml` and route them to a different URL.
+
+```toml title="pixi.toml"
+channels = ["conda-forge", "bioconda", "private-channel"]
+
+[[channels]] # (1)!
````
Is this structure possible in TOML?

wolfv commented:
OK, I think I also gathered my thoughts. 

I think what we need to define is that `conda.anaconda.org/${channel_name}` is currently the "default namespace" for channels. Companies or individuals might want to change this default namespace to something else (e.g. `repo.prefix.dev` or some internal mirror).

The default namespace means that any named channel is resolved using that URL.

It also means that in the lockfile we store the resolved namespace + channel as URL to the files.

If we configure mirrors for a given channel like:

```toml
[[channels]]
name = "bioconda"
mirrors = [
   "https://internal-mirror.com/bioconda"
]
```

We still store "canonical" (resolved) name to the namespace in the lockfile (`conda.anaconda.org/bioconda`) which will help with the portability of the lockfile. It also means that a mirror should always (ideally) be an identical copy of the canonical source.

Now you may also want to configure some channel alias or _not_ use the default namespace for a given channel. For example:

```
[[channels]]
name = "prefix-internal"
url = "https://repo.prefix.dev/prefix-internal"
```

In this case, we ignore the default channel namespace and instead use the `url` directly. The lockfile would now also refer to the `https://repo.prefix.dev/...` url and not the default namespace url. This configuration is basically just a shorthand instead of inserting the full URL as the channel in the `pixi.toml`.

I understand why `url` and `mirrors` might be confusing initially but I do think that they have two somewhat distinct uses. Mirrors should just always be mirrors of the canonical source, while URLs are more like aliases.

Does this explanation change your way of thinking a bit @pavelzw / @0xbe7a ?

0xbe7a commented:
I think the "URL" field allows channel definitions outside the "pixi.toml", which means that lockfiles are not reproducible without your global channel configuration.

wolfv commented:
@0xbe7a the lockfile would be reproducible because we would store the entire URL as part of the lockfile. However, reproducing the lockfile from the `pixi.toml` would require the same global setting, that is correct.

pavelzw commented:
> we would store the entire URL as part of the lockfile

I personally would like to specify this in pixi.toml s.t. `rm pixi.lock && pixi install` generates the same lockfile no matter whether you have the rattler.toml or not.

wolfv commented:
OK, then maybe it shoudl be like the following:

```toml
# these two value change the `pixi.toml` generated by init (if channels are not otherwise overriden)
default_channels = ["conda-forge", "bioconda"]
default_server = ["https://repo.prefix.dev"]

# mirror config is global and used at runtime but does not change the lockfile at all
[mirrors]
"https://conda.anaconda.org" = [
   "https://conda.anaconda.org", # need to re-add this to also use anaconda.org
   "https://repo.prefix.dev",
   "oci://ghcr.io/channel-mirrors",
   { url = "https://mirrorme.com", no_jlap = true }
]
"https://conda.anaconda.org/conda-forge" = [
   "https://internal-conda-forge.com/"
]
```

wolfv commented:
In the `pixi.toml` we would add a line that looks like:

```
server = "https://repo.prefix.dev"
channels = ["conda-forge", "bioconda"]
```

Although the exact naming of this is not defined yet. Should it be `channel_server` or `channel_namespace`?.

Also to note is that this would influence the mirror selection. If you still want to use a different mirror when checking for conda-forge you would now have to add a new mirror line that specifies the mirrors for `https://repo.prefix.dev`.

With the proposed config we are very specific about the functionality and it's quite straightforward to understand. We could implement simplificiations later on. 

I also looked at dockers `registry-mirrors` config: https://docs.docker.com/docker-hub/mirror/#configure-the-docker-daemon

And in `buildx`: https://docs.docker.com/build/buildkit/configure/#registry-mirror

Which look kinda similar :)

ruben-arts commented:
This is not up to date, do we want this in the documentation?

pavelzw commented:
I don't have strong opinions on this. I'm also happy to close this when the docs in #988 are fully implemented 🤷🏻

wolfv commented:
Let's close this one as we've merged most changes. We should chekc if the new options work well for all users and we can revisit this :)

We still need to implement the `default_server` and `server` options for pixi, though (I'll make a separate issue for that).
